### PR TITLE
feat: added support for multi pem block files for all plugins. added support for pkcs#1 decryption

### PR DESCRIPTION
**Pull request by bustedware**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

### Notes on deprecation of pkcs#1 routines
[golang issue](https://github.com/golang/go/issues/59961) raised in response to deprecation
[comments on pr](https://go-review.googlesource.com/c/go/+/400655/2#message-0fd004992a19b389e27219f5a8dc5cd6a06f5bb7) associated to fixing deprecation

in the comments
> Note that there is no plan to actually remove this functions. That would break the Go 1 compatibility guarantee. They are deprecated to encourage people to use other functionality in new code.

we should still be able to decrypt pkcs#1 pem blocks since there is no longer a concern with x509.IsEncryptedPEMBlock() and x509.DecryptPEMBlock() going away in a future golang release. i believe there was a concern about it [boomeranging](https://influxcommunity.slack.com/archives/C019JDRJAE7/p1635755145025200?thread_ts=1635478730.019900&cid=C019JDRJAE7) back and needing removed but since its clear golang will keep the method available that means there must enough demand for backwards compatibility. [related PR](https://github.com/influxdata/telegraf/pull/10020) which initially introduced this feature. 

choosing between pkcs#1 or pkcs#8 should be up to the consumers of telegraf, not the contributors

### CODE REVIEW

**Comment by powersj** [comment]
@jdstrand can I request a review from you please? thanks!

File: plugins/common/tls/utils.go [hunk 1]
```diff
@@ -36,3 +43,58 @@ func ParseTLSVersion(version string) (uint16, error) {
 	sort.Strings(available)
 	return 0, fmt.Errorf("unsupported version %q (available: %s)", version, strings.Join(available, ","))
 }
+
+func readFile(filename string) []byte {
+	octets, err := os.ReadFile(filename)
+	if err != nil {
+		panic(fmt.Sprintf("reading %q: %v", filename, err))
```
Comment (Hipska) [inline plugins/common/tls/utils.go:50]: I don't think we should introduce panics, same for occurrences below..

**Review by Hipska — CHANGES_REQUESTED** [review]

File: plugins/common/tls/utils.go [hunk 2]
```diff
@@ -36,3 +43,58 @@ func ParseTLSVersion(version string) (uint16, error) {
...
+	}
+	return octets
+}
+
+func ReadCertificate(filename string) string {
+	octets := readFile(filename)
+	return string(octets)
+}
+
+func ReadKey(filename string, password string) string {
```
Comment (jdstrand) [inline plugins/common/tls/utils.go:60]: We're returning `string` here only to have it be cast back to `[]byte` in `loadCertificate()` when it calls `tls.X509KeyPair()`. I think it would be best to just return `[]byte` here. We should also be returning `error` here instead of panicking below, and handling that in `loadCertificate()`.

File: plugins/common/tls/utils.go [hunk 3]
```diff
@@ -36,3 +43,58 @@ func ParseTLSVersion(version string) (uint16, error) {
...
+		} else {
+			pemBlock := string(pem.EncodeToMemory(&pem.Block{Type: currentBlock.Type, Bytes: currentBlock.Bytes}))
+			allBlocks += pemBlock
+		}
+		currentBlock, remainingBlocks = pem.Decode(remainingBlocks)
+		if currentBlock == nil {
+			break
+		}
+	}
+	return allBlocks
```
Comment (jdstrand) [inline plugins/common/tls/utils.go:99]: IIUC, `allblocks` will return concatenated strings of multiple blocks, which I think is wrong since a) with a combined file, `allblocks` will have both the private key and the cert in it, which would be unexpected with the second argument of `tls.X509KeyPair()` and b) `allblocks` will contain multiple private keys if the key file has them, which is also unexpected with `tls.X509KeyPair()`. https://pkg.go.dev/crypto/tls#X509KeyPair suggests that the second argument should only be a single private block. I suggest not relying on implementation dependent behavior in `tls.X509KeyPair()` and make this code only return one private block.

Since this is called `ReadKey()`, I think it makes sense to:
* ignore non-private keys instead of appending them
* return the first private key rather than appending them (with a code comment)

File: plugins/common/tls/utils.go [hunk 4]
```diff
@@ -36,3 +43,58 @@ func ParseTLSVersion(version string) (uint16, error) {
...
+			}
+			var decryptedKey *rsa.PrivateKey
+			decryptedKey, err := pkcs8.ParsePKCS8PrivateKeyRSA(currentBlock.Bytes, []byte(password))
+			if err != nil {
+				panic(fmt.Errorf("failed to parse encrypted PKCS#8 private key: %w", err))
+			}
+			pemBlock := string(pem.EncodeToMemory(&pem.Block{Type: currentBlock.Type, Bytes: x509.MarshalPKCS1PrivateKey(decryptedKey)}))
+			allBlocks += pemBlock
+		} else if currentBlock.Headers["Proc-Type"] == "4,ENCRYPTED" {
+			decryptedKeyDER, err := x509.DecryptPEMBlock(currentBlock, []byte(password))
```
Comment (jdstrand) [inline plugins/common/tls/utils.go:80]: Should we also error here if no password like we do in the previous stanza?

File: plugins/common/tls/utils.go [hunk 5]
```diff
@@ -36,3 +43,58 @@ func ParseTLSVersion(version string) (uint16, error) {
...
+
+func readFile(filename string) []byte {
+	octets, err := os.ReadFile(filename)
+	if err != nil {
+		panic(fmt.Sprintf("reading %q: %v", filename, err))
+	}
+	return octets
+}
+
+func ReadCertificate(filename string) string {
```
Comment (jdstrand) [inline plugins/common/tls/utils.go:55]: We're returning `string` here only to cast it to `[]byte` in `loadCertificate()` when it calls `tls.X509KeyPair()`. I think it would be best to just return []byte here. We should also be returning error here and handling that in loadCertificate().

File: plugins/common/tls/utils.go [hunk 6]
```diff
@@ -36,3 +43,58 @@ func ParseTLSVersion(version string) (uint16, error) {
...
+	octets := readFile(filename)
```
Comment (jdstrand) [inline plugins/common/tls/utils.go:56]: AIUI, this will end up feeding both the public and private keys to the first argument of `tls.X509KeyPair()` or multiple blocks if more than one certificate and/or private key is in the combined file. https://pkg.go.dev/crypto/tls#X509KeyPair suggests that this should only be a single cert block. I suggest not relying on implementation dependent behavior in `tls.X509KeyPair()` and make this code only return one cert block.

**Review by jdstrand — CHANGES_REQUESTED** [review]
Thanks for the PR. I found it a bit hard to follow the history with all the merges to master/etc. The changes we care about seem to be in these commits:
* 1161eaa8ba8b9e6e6d94d97fc7cf51f4dadfc3b2
* cb738bd72c22ca6ace4061e8ac0ed0110fb1c577
* 2c1e0cfe20e3fdb3b05570879163295b88b9ca88
* 2ea45492e8186a0f67556513142aea9b06d6fa46
* e57b199b3bc647deebef3a2fa3b1c8d91f03338c

Personally, it would be nice to see in the PR description the breakout of the different commits on top of a rebased master (at least before people start reviewing the PR). I'm not the maintainer of telegraf so I defer to @powersj on if this would be helpful to him.

Overall, I think its fine to add back this functionality, but I have a number of suggestions inline.
